document behavior of per handler read/write timeout with test(s)

tkashem opened the pull request:
#### What type of PR is this?

/kind feature


#### What this PR does / why we need it:

In support of https://github.com/kubernetes/kubernetes/pull/114189
This PR documents the behavior of per request read/write handler with net/http server

#### Which issue(s) this PR fixes:
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/4462
```

tkashem commented:
/assign @wojtek-t @sttts 
/cc @MikeSpreitzer 

(I will add the result of running stress here soon)

tkashem commented:
```
$ go test -race k8s.io/apiserver/pkg/endpoints/filters -c

$ stress ./filters.test -test.run=TestPerHandlerWriteDeadlineHTTP2WithWriteShouldReturnErrorAfterTimeout -test.cpu=32
5s: 24 runs so far, 0 failures
10s: 56 runs so far, 0 failures
15s: 88 runs so far, 0 failures
...
5m15s: 1977 runs so far, 0 failures
5m20s: 2009 runs so far, 0 failures


$ stress ./filters.test -test.run=TestPerHandlerWriteDeadlineHTTP1WithWriteShouldReturnErrorAfterTimeout -test.cpu=32
5s: 0 runs so far, 0 failures
10s: 0 runs so far, 0 failures
15s: 8 runs so far, 0 failures
...
10m45s: 456 runs so far, 0 failures
10m50s: 464 runs so far, 0 failures
10m55s: 464 runs so far, 0 failures


$ stress ./filters.test -test.run=TestPerHandlerWriteDeadlineHTTP2WithTimeoutBeforeHandlerWrites -test.cpu=32
5s: 24 runs so far, 0 failures
10s: 56 runs so far, 0 failures
15s: 88 runs so far, 0 failures
...
5m30s: 2088 runs so far, 0 failures
5m35s: 2123 runs so far, 0 failures
5m40s: 2155 runs so far, 0 failures


$ stress -p=32 ./filters.test -test.run=TestPerHandlerWriteDeadlineHTTP1WithTimeoutBeforeHandlerWrites -test.cpu=64
5s: 0 runs so far, 0 failures
10s: 0 runs so far, 0 failures
15s: 32 runs so far, 0 failures
20s: 32 runs so far, 0 failures
25s: 64 runs so far, 0 failures
...
5m55s: 992 runs so far, 0 failures
6m0s: 1024 runs so far, 0 failures
6m5s: 1024 runs so far, 0 failures


$ stress -p=32 ./filters.test -test.run=TestPerHandlerWriteDeadlineHTTP2WithTimeoutAfterHandlerWrites -test.cpu=64
5s: 96 runs so far, 0 failures
10s: 224 runs so far, 0 failures
15s: 352 runs so far, 0 failures
20s: 469 runs so far, 0 failures
...
5m40s: 8482 runs so far, 0 failures
5m45s: 8610 runs so far, 0 failures
5m50s: 8728 runs so far, 0 failures
5m55s: 8847 runs so far, 0 failures


$ stress -p=32 ./filters.test -test.run=TestPerHandlerWriteDeadlineHTTP1WithTimeoutAfterHandlerWrites -test.cpu=64
5s: 96 runs so far, 0 failures
10s: 224 runs so far, 0 failures
15s: 352 runs so far, 0 failures
20s: 474 runs so far, 0 failures
...
6m30s: 9635 runs so far, 0 failures
6m35s: 9759 runs so far, 0 failures
6m40s: 9881 runs so far, 0 failures


$ stress -p=32 ./filters.test -test.run=TestPerHandlerWriteTimeoutHTTP2WithHandlerWritingIndefinitely -test.cpu=64
5s: 70 runs so far, 0 failures
10s: 157 runs so far, 0 failures
15s: 239 runs so far, 0 failures
...
6m5s: 6317 runs so far, 0 failures
6m10s: 6406 runs so far, 0 failures
6m15s: 6492 runs so far, 0 failures
6m20s: 6577 runs so far, 0 failures


$ stress -p=32 ./filters.test -test.run=TestPerHandlerReadDeadlineHTTP2WithClientNotWritingToRequestBody -test.cpu=64
5s: 96 runs so far, 0 failures
10s: 224 runs so far, 0 failures
15s: 352 runs so far, 0 failures
20s: 480 runs so far, 0 failures
...
5m50s: 8641 runs so far, 0 failures
5m55s: 8769 runs so far, 0 failures
6m0s: 8897 runs so far, 0 failures
6m5s: 9015 runs so far, 0 failures


$ stress -p=32 ./filters.test -test.run=TestPerHandlerReadDeadlineHTTP1WithClientNotWritingToRequestBody -test.cpu=64
5s: 96 runs so far, 0 failures
10s: 224 runs so far, 0 failures
15s: 352 runs so far, 0 failures
20s: 480 runs so far, 0 failures
...
5m5s: 7570 runs so far, 0 failures
5m10s: 7691 runs so far, 0 failures
5m15s: 7819 runs so far, 0 failures
5m20s: 7939 runs so far, 0 failures


$ stress -p=32 ./filters.test -test.run=TestPerHandlerReadDeadlineHTTPWithTimeoutWhileClientIsSendingContent -test.cpu=64
5s: 96 runs so far, 0 failures
10s: 227 runs so far, 0 failures
15s: 350 runs so far, 0 failures
20s: 471 runs so far, 0 failures
...
4m55s: 7381 runs so far, 0 failures
5m0s: 7508 runs so far, 0 failures
5m5s: 7636 runs so far, 0 failures
5m10s: 7764 runs so far, 0 failures


$ stress -p=32 ./filters.test -test.run=TestPerHandlerReadDeadlineHTTP2WithNoRequestBody -test.cpu=64
5s: 119 runs so far, 0 failures
10s: 256 runs so far, 0 failures
15s: 402 runs so far, 0 failures
20s: 535 runs so far, 0 failures
...


```

jiahuif commented:
/triage accepted

tkashem commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/request_deadline_test.go:
```diff
@@ -438,6 +443,961 @@ func TestWithFailedRequestAudit(t *testing.T) {
+
+		// NOTE: if we block the handler here, the client (not setting a
+		// transport or client timeout) will hang indefinitely, even
+		// though we have set the write deadline on the server side.
```
self: open an issue with the golang team

MikeSpreitzer replied:
The description at https://pkg.go.dev/net/http@go1.21.6#ResponseController.SetWriteDeadline is consistent with the behavior you report: the deadline is not a deadline for some (something, anything), it is _only_ a tripwire for write operations that happen after the deadline.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/request_deadline_test.go:
```diff
@@ -438,6 +443,961 @@ func TestWithFailedRequestAudit(t *testing.T) {
+
+		ctrl := http.NewResponseController(w)
+		if err := ctrl.SetWriteDeadline(time.Now().Add(100 * time.Millisecond)); err != nil {
+			t.Errorf("expected no error from SetReadDeadline, but got: %v", err)
```
Typo here, "Read" vs "Write"?

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/request_deadline_test.go:
```diff
@@ -438,6 +443,961 @@ func TestWithFailedRequestAudit(t *testing.T) {
+
+		ctrl := http.NewResponseController(w)
+		if err := ctrl.SetWriteDeadline(time.Now().Add(5 * time.Millisecond)); err != nil {
+			t.Errorf("expected no error from SetReadDeadline, but got: %v", err)
```
typo here?

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/request_deadline_test.go:
```diff
@@ -438,6 +443,961 @@ func TestWithFailedRequestAudit(t *testing.T) {
+		}
+
+		// wait for timeout to occur, 10s to eliminate flakes in ci
+		<-time.After(10 * time.Second)
```
Why not wait here like TestPerHandlerWriteDeadlineHTTP2WithWriteShouldReturnErrorAfterTimeout does?

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/request_deadline_test.go:
```diff
@@ -438,6 +443,961 @@ func TestWithFailedRequestAudit(t *testing.T) {
+	}
+}
+
+func TestPerHandlerWriteDeadlineHTTP1WithWriteShouldReturnErrorAfterTimeout(t *testing.T) {
```
This function is very similar to TestPerHandlerWriteDeadlineHTTP2WithWriteShouldReturnErrorAfterTimeout ; I think it would be better if this were more DRY.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/request_deadline_test.go:
```diff
@@ -438,6 +443,961 @@ func TestWithFailedRequestAudit(t *testing.T) {
+
+		ctrl := http.NewResponseController(w)
+		if err := ctrl.SetWriteDeadline(time.Now().Add(100 * time.Millisecond)); err != nil {
+			t.Errorf("expected no error from SetReadDeadline, but got: %v", err)
```
typo

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/request_deadline_test.go:
```diff
@@ -438,6 +443,961 @@ func TestWithFailedRequestAudit(t *testing.T) {
+	}
+}
+
+func TestPerHandlerWriteTimeoutHTTP2WithHandlerWritingIndefinitely(t *testing.T) {
```
Why no HTTP/1 variant of this one?

MikeSpreitzer commented:
I see a lot of code duplicated between the HTTP/1 and HTTP/2 variants of each test, in only one test is that duplication avoided. I think the duplication should be avoided in all the tests.  Further, there is a lot of duplication between the progression of the write tests, it _might_ be good to dedup that too.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1134 @@
+			waiter: &waitWithDuration{after: 5 * time.Second},
+			verifyClientErrFn: func(t *testing.T, err error) {
+				t.Helper()
+				// TODO: look into why we see this error
```
is this a blocker for the PR ?

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1134 @@
+			// 5s should be enough to withstand flakes in CI.
+			waiter: &waitWithDuration{after: 5 * time.Second},
+			verifyClientErrFn: func(t *testing.T, err error) {
+				t.Helper()
```
this hides the line of the failing  `t.Error`. Do we want that?

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1134 @@
+		protoMajor        int
+		waiter            waiter
+		verifyClientErrFn func(*testing.T, error)
+	}{
```
a `name string` would help to understand what these are testing.

sttts replied:
much better

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1134 @@
+			server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+				defer close(handlerDoneCh)
+				if req.ProtoMajor != test.protoMajor {
+					t.Errorf("expected ProtoMajor: %d, but got: %s", test.protoMajor, req.Proto)
```
this is another go routine, isn't it? I believe you are not allowed to call `t.Errorf`.

tkashem replied:
I thought calling `t.Errorf` is safe as long as it is called before the test completes, but i am seeing a data race when the test uses `t.Parallel()` and has sub tests.  I will investigate the data race issue I am seeing, for now i am calling `t.Errorf` from within the test.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1134 @@
+					}
+				}()
+			}))
+
```
some `t.Logf("Doing X")` would make this test easier to read.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/testing/recorder.go:
```diff
@@ -0,0 +1,31 @@
+	// we get the the following stream reset error due to a
+	// timeout from the per handler write timeout.
+	return strings.Contains(err.Error(), "stream error: stream ID") &&
+		strings.Contains(err.Error(), "INTERNAL_ERROR")
```
helpful to add a small unit test to verify that these error strings don't change?

tkashem replied:
yes, added

sttts commented:
General comment: these tests are hard to read and it is not clear for the read which parts are boilerplate and which are actually assertions. Some `t.Log` for documentation and proper human-readable text descriptions (`map[string]test` or test { name string; ... }`) would help readibility.

tkashem commented:
> I see a lot of code duplicated between the HTTP/1 and HTTP/2 variants of each test, in only one test is that duplication avoided. I think the duplication should be avoided in all the tests. Further, there is a lot of duplication in the progression of the write tests, it _might_ be good to dedup that too.

@MikeSpreitzer for each use case, i consolidated both http/1x and http/2.0. I would like to keep each use case in a separate test so it is easier to read, thoughts?

tkashem commented:
/hold

i can reproduce a data race when `t.Logf` is used from a goroutine, I am going to investigate the issue, otherwise, please continue to review the tests

```
$ stress -p=32 ./filters.test -test.run=TestPer
5s: 0 runs so far, 0 failures
10s: 0 runs so far, 0 failures

5m40s: 911 runs so far, 0 failures

/tmp/go-stress-20240503T111346-354534958
==================
WARNING: DATA RACE
Read at 0x00c000356523 by goroutine 1108:
  testing.(*common).logDepth()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/testing/testing.go:1023 +0x504
  testing.(*common).log()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/testing/testing.go:1010 +0xa4
  testing.(*common).Logf()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/testing/testing.go:1061 +0x6a
  k8s.io/apiserver/pkg/endpoints/filters.TestPerRequestWriteDeadlineWithWriteShouldReturnErrorAfterTimeout.func1.1.1.1()
      /home/akashem/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:97 +0x12f
  runtime.deferreturn()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/runtime/panic.go:602 +0x5d
  k8s.io/apiserver/pkg/endpoints/filters.TestPerRequestWriteDeadlineWithWriteShouldReturnErrorAfterTimeout.func1.1()
      /home/akashem/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:107 +0x274
  net/http.HandlerFunc.ServeHTTP()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:2166 +0x47
  net/http.serverHandler.ServeHTTP()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:3137 +0x2a1
  net/http.initALPNRequest.ServeHTTP()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:3745 +0x35e
  net/http.(*initALPNRequest).ServeHTTP()
      <autogenerated>:1 +0x7b
  net/http.Handler.ServeHTTP-fm()
      <autogenerated>:1 +0x67
  net/http.(*http2serverConn).runHandler()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/h2_bundle.go:6153 +0x1f3
  net/http.(*http2serverConn).scheduleHandler.gowrap1()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/h2_bundle.go:6088 +0x5d

Previous write at 0x00c000356523 by goroutine 56:
  testing.tRunner.func1()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/testing/testing.go:1676 +0x8fa
  runtime.deferreturn()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/runtime/panic.go:602 +0x5d
  testing.(*T).Run.gowrap1()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/testing/testing.go:1742 +0x44

Goroutine 1108 (running) created at:
  net/http.(*http2serverConn).scheduleHandler()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/h2_bundle.go:6088 +0x28a
  net/http.(*http2serverConn).processHeaders()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/h2_bundle.go:5810 +0xd0e
  net/http.(*http2serverConn).processFrame()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/h2_bundle.go:5312 +0x40a
  net/http.(*http2serverConn).processFrameFromReader()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/h2_bundle.go:5255 +0x2dc
  net/http.(*http2serverConn).serve()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/h2_bundle.go:4742 +0x16fe
  net/http.(*http2Server).ServeConn()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/h2_bundle.go:4311 +0x1a90
  net/http.http2ConfigureServer.func1()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/h2_bundle.go:4101 +0x1a9
  net/http.(*conn).serve()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:1947 +0xd41
  net/http.(*Server).Serve.gowrap3()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:3285 +0x4f

Goroutine 56 (finished) created at:
  testing.(*T).Run()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/testing/testing.go:1742 +0x825
  testing.runTests.func1()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/testing/testing.go:2161 +0x85
  testing.tRunner()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/testing/testing.go:1689 +0x21e
  testing.runTests()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/testing/testing.go:2159 +0x8be
  testing.(*M).Run()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/testing/testing.go:2027 +0xf17
  main.main()
      _testmain.go:141 +0x2bd
==================
--- FAIL: TestPerRequestWriteDeadlineWithTimeoutOccursAfterHandlerFlushes (5.23s)
    --- FAIL: TestPerRequestWriteDeadlineWithTimeoutOccursAfterHandlerFlushes/HTTP/1 (5.09s)
        testing.go:1398: race detected during execution of test
--- FAIL: TestPerRequestWriteDeadlineWithSlowReader (4.92s)
    read_write_deadline_test.go:892: server: the handler completed, written 541kB of payload, took: 50.180339ms
    read_write_deadline_test.go:1011: elapsed=1.117136497s, read=125kB
    read_write_deadline_test.go:1011: elapsed=2.227730816s, read=250kB
    read_write_deadline_test.go:1011: elapsed=3.353076487s, read=375kB
    read_write_deadline_test.go:1011: elapsed=4.539827169s, read=500kB
    read_write_deadline_test.go:1014: elapsed=4.890938746s, read=540KB
    testing.go:1398: race detected during execution of test
--- FAIL: TestPerRequestReadDeadlineWithNoRequestBody (10.19s)
    --- FAIL: TestPerRequestReadDeadlineWithNoRequestBody/HTTP/2 (5.04s)
        read_write_deadline_test.go:828: request Body type: *http.http2requestBody
        testing.go:1398: race detected during execution of test
FAIL

```

sttts commented:
@tkashem yes, reads much better. Alternative (if volume is small): t.Log. Give nicer logs when something goes wrong.

tkashem commented:
/hold cancel

(I fixed the data race mentioned at https://github.com/kubernetes/kubernetes/pull/122923#issuecomment-2093239301)

tkashem commented:
/retest

tkashem commented:
/test pull-kubernetes-e2e-kind-cloud-provider-loadbalancer

tkashem committed:
add test to document behavior of net/http read/write deadline

tkashem commented:
sharing `stress` results

```
$ go test -race k8s.io/apiserver/pkg/endpoints/filters -c
```

```
$ stress -p=32 ./filters.test -test.run=TestPerRequestWriteDeadlineWithWriteShouldReturnErrorAfterTimeout
5s: 0 runs so far, 0 failures
10s: 32 runs so far, 0 failures
15s: 64 runs so far, 0 failures
...
```

tkashem commented:
sharing `stress` results

```
$ go test -race k8s.io/apiserver/pkg/endpoints/filters -c
```

```
$ stress -p=256 ./filters.test -test.run=TestPerRequestWriteDeadlineWithWriteShouldReturnErrorAfterTimeout
5s: 0 runs so far, 0 failures
10s: 256 runs so far, 0 failures
15s: 330 runs so far, 0 failures
...
5m10s: 11939 runs so far, 0 failures
5m15s: 12175 runs so far, 0 failures
5m20s: 12395 runs so far, 0 failures
```

```
$ stress -p=256 ./filters.test -test.run=TestPerRequestWriteDeadlineWithTimeoutOccursBeforeHandlerWrites
5s: 0 runs so far, 0 failures
10s: 256 runs so far, 0 failures
15s: 323 runs so far, 0 failures
...
4m55s: 11658 runs so far, 0 failures
5m0s: 11876 runs so far, 0 failures
5m5s: 12081 runs so far, 0 failures
```

```
$ stress -p=256 ./filters.test -test.run=TestPerRequestWriteDeadlineWithTimeoutOccursAfterHandlerWrites
5s: 0 runs so far, 0 failures
10s: 256 runs so far, 0 failures
15s: 338 runs so far, 0 failures
...
4m50s: 11484 runs so far, 0 failures
4m55s: 11713 runs so far, 0 failures
5m0s: 11931 runs so far, 0 failures
```

```
$ stress -p=256 ./filters.test -test.run=TestPerRequestWriteDeadlineWithTimeoutOccursAfterHandlerFlushes
5s: 0 runs so far, 0 failures
10s: 256 runs so far, 0 failures
15s: 352 runs so far, 0 failures
...
5m10s: 12316 runs so far, 0 failures
5m15s: 12536 runs so far, 0 failures
5m20s: 12750 runs so far, 0 failures
```

```
$ stress -p=256 ./filters.test -test.run=TestPerRequestWriteDeadlineWithHandlerWritingIndefinitely
5s: 11 runs so far, 0 failures
10s: 294 runs so far, 0 failures
15s: 545 runs so far, 0 failures
...
4m50s: 13214 runs so far, 0 failures
4m55s: 13442 runs so far, 0 failures
5m0s: 13674 runs so far, 0 failures
```

```
$ stress -p=64 ./filters.test -test.run=TestPerRequestReadDeadlineWithClientNotWritingToRequestBody
5s: 86 runs so far, 0 failures
10s: 233 runs so far, 0 failures
15s: 384 runs so far, 0 failures
...
6m45s: 11444 runs so far, 0 failures
6m50s: 11586 runs so far, 0 failures
6m55s: 11728 runs so far, 0 failures
```

```
$ stress -p=64 ./filters.test -test.run=TestPerRequestReadDeadlineWithTimeoutOccursWhileClientIsSendingContent
5s: 102 runs so far, 0 failures
10s: 263 runs so far, 0 failures
15s: 422 runs so far, 0 failures
...
5m50s: 10215 runs so far, 0 failures
5m55s: 10358 runs so far, 0 failures
6m0s: 10503 runs so far, 0 failures
```

```
$ stress -p=256 ./filters.test -test.run=TestPerRequestReadDeadlineWithNoRequestBody
5s: 0 runs so far, 0 failures
10s: 0 runs so far, 0 failures
15s: 256 runs so far, 0 failures
...
5m0s: 6656 runs so far, 0 failures
5m5s: 6789 runs so far, 0 failures
5m10s: 6912 runs so far, 0 failures
```

```
$ stress -p=256 ./filters.test -test.run=TestPerRequestReadWriteDeadlineWithHijack
5s: 0 runs so far, 0 failures
10s: 256 runs so far, 0 failures
15s: 317 runs so far, 0 failures
...
4m55s: 11791 runs so far, 0 failures
5m0s: 12025 runs so far, 0 failures
5m5s: 12231 runs so far, 0 failures
```

```
$ stress -p=256 ./filters.test -test.run=TestPerRequestWriteDeadlineWithConnectionReuse
5s: 0 runs so far, 0 failures
10s: 256 runs so far, 0 failures
15s: 343 runs so far, 0 failures
...
5m0s: 11552 runs so far, 0 failures
5m5s: 11753 runs so far, 0 failures
5m10s: 11908 runs so far, 0 failures
```

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1459 @@
+			// write timeout is set to 100ms, a wait of 5s should be
+			// enough to withstand flakes in CI.
+			waiter:             &waitWithDuration{after: 5 * time.Second},
+			clientErrVerifier:  wantContains{contains: "local error: tls: bad record MAC"},
```
won't these break with the next Golang update? They look very low-level and might be changed at any time.

MikeSpreitzer replied:
Are you aware of a specific change coming? I just tested this with go version 1.23rc2, and got no failure.

Does go provide a more abstract way to test for this condition? Absent that, what would you recommend testing for here?

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1459 @@
+func (w waitWithDuration) wait() <-chan time.Time { return time.After(w.after) }
+func (w waitWithDuration) close()                 {}
+
+type waitWithChannelClose struct {
```
nit: waitForClose

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1459 @@
+}
+
+func (w waitWithChannelClose) wait() <-chan time.Time {
+	// for http/2, we do the following:
```
this comment does not make sense here. This object has nothing to do with http/2.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1459 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+func TestPerRequestWriteDeadlineWithWriteShouldReturnErrorAfterTimeout(t *testing.T) {
```
TestWriteErrorAfterRequestDeadline

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1459 @@
+				<-test.waiter.wait()
+
+				// c) keep writing 1kB of data at a time until
+				// Write returns an error
```
eventually

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1459 @@
+				// c) write a few bytes so there is something in the buffer to
+				// flush, the client should never see these bytes in the response.
+				if _, err := w.Write([]byte("hello")); err != nil {
+					t.Errorf("server: unexpected error from Write after timeout: %v", err)
```
why is this guaranteed to succeed and the 1kb writes above are not? Is Golang specifying some buffer size?

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1459 @@
+	}
+}
+
+func TestPerRequestWriteDeadlineWithTimeoutOccursBeforeHandlerWrites(t *testing.T) {
```
TestImmediateFlushErrorAfterRequestDeadline

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1459 @@
+						t.Logf("server: after timeout, Write wrote %dkB, duration: %s", written/1000, time.Since(now))
+					}()
+					for {
+						n, err := w.Write(bytes.Repeat([]byte("a"), 1000))
```
shouldn't we check that none of these makes it to the client? You claim something below before flushing.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1459 @@
+	}
+}
+
+func TestPerRequestWriteDeadlineWithTimeoutOccursAfterHandlerWrites(t *testing.T) {
```
TestFlushErrorAfterRequestDeadline

In general, don't describe just the preconditions, but describe the claimed behaviour in test names.

tkashem replied:
renamed with claimed behavior, I also want to keep the prefix `TestPerRequestWith` so it is easier to run these tests in a group.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1459 @@
+				<-test.waiter.wait()
+
+				// e) write message 2 (the client should never
+				//    see the following message)
```
isn't this the actual interesting bit? Flush failing, sure. But that nothing is sent anymore.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1459 @@
+	}
+}
+
+func TestPerRequestWriteDeadlineWithTimeoutOccursAfterHandlerFlushes(t *testing.T) {
```
TestNothingReceivedFromAfterRequestDeadline

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1459 @@
+			func() {
+				defer test.waiter.close()
+
+				_, err := server.Client().Get(server.URL)
```
What surprises is that you get an error, but there has been data written before. So wouldn't I expect a body? And then wouldn't I have to close that body? I.e. this is leaking.

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1459 @@
+				}
+
+				// b) keep writing to the ResponseWriter indefinitely
+				if _, err := io.Copy(w, neverEnding('a')); err != nil {
```
how is this different to the 1kb write loop in the first test?

sttts commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1459 @@
+	//       an "i/o timeout" error
+	t.Parallel()
+
+	const deadline = 500 * time.Millisecond
```
why all these different timeout in the tests?

sttts commented:
/lgtm
/approve

Nothing blocking, couple of nits and questions and style.

/hold

Unhold at will or fix some of the nits.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go:
```diff
@@ -0,0 +1,1459 @@
+	}{
+		{
+			protoMajor: 1, // http/1x
+			// write timeout is set to 100ms, a wait of 5s should be
+			// enough to withstand flakes in CI.
+			waiter:             &waitWithDuration{after: 5 * time.Second},
```
Why use dead reckoning here instead of the coordination via channels as in the HTTP/2 case?

tkashem replied:
using a channel based waiter with http/1x blocks the handler indefinitely, since the request handler is executed in the same goroutine as the connection serving goroutine, the KEP explains this: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4460-per-request-deadline#client-hanging-indefinitely

tkashem committed:
fixup! add test to document behavior of net/http read/write deadline

tkashem commented:
These new tests run in parallel, and add `11s` to test time for this package
```
go test -count=1 -race k8s.io/apiserver/pkg/endpoints/filters -run=TestPerRequest
ok  	k8s.io/apiserver/pkg/endpoints/filters	11.157s
```

tkashem commented:
new `stress` result:
```
$ stress -p=512 ./filters.test -test.run=TestPerRequest -test.parallel=16
5s: 0 runs so far, 0 failures
10s: 0 runs so far, 0 failures
15s: 61 runs so far, 0 failures
...
7m5s: 7528 runs so far, 0 failures
7m10s: 7624 runs so far, 0 failures
7m15s: 7689 runs so far, 0 failures
```

tkashem commented:
@sttts I addressed the feedback in this fix up commit: https://github.com/kubernetes/kubernetes/pull/122923/commits/f91cdf768dcd893e2d18a5705071b899f979374d

MikeSpreitzer commented:
/retest

tkashem commented:
/retest

sttts commented:
/lgtm
/approve

sttts commented:
/hold cancel
